### PR TITLE
Sync Dao-AILab FlashAttention through ce088ab9

### PR DESCRIPTION
**Pull request by StevenWang-CY**
## Summary

This sync imports the applicable non-FA3 changes from the 26 Dao-AILab commits after the cutoff used by #149 and advances the fork through Dao-AILab commit `ce088ab9`. The vLLM fork's FlashAttention 3 implementation under `hopper/` remains byte-for-byte frozen at current fork `main`, per maintainer direction.

| Reference | Commit |
| --- | --- |
| Previous Dao-AILab cutoff | `5835c733e7e9c07606b045255768e8a7e9e851bd` |
| Intermediate cutoff prepared in #171 | `2409214a03797b168f648ea30df1adbc09ce658a` |
| vLLM fork base | `06bdd47c0d0383daf6a2ff0c418faff9c6da16e5` |
| Current fork `main` merged into the candidate | `88a8b4c9980b7060ac249742b4be8dd8a2769e9d` |
| New Dao-AILab cutoff | `ce088ab9ce0fc0434dcd8afa0a791da9fcc3a820` |

The first six Dao-AILab commits were adapted in #171. This branch carries those resolutions onto the current fork tree, replays the twenty later commits, and preserves each upstream human author. It also retains #171's revert of #166 because the upstream FP8 correction in Dao-AILab #2717 supersedes that threshold workaround.

## Upstream changes

- Architecture and kernel coverage: SM103 `tcgen.ld.red` support ([Dao-AILab #2696](https://github.com/Dao-AILab/flash-attention/pull/2696)); 2CTA SM100 block-sparse backward ([#2661](https://github.com/Dao-AILab/flash-attention/pull/2661)); score modifiers in variable-length backward ([#2547](https://github.com/Dao-AILab/flash-attention/pull/2547)); learnable-sink backward on SM90, SM100, and SM110 ([#2706](https://github.com/Dao-AILab/flash-attention/pull/2706)).
- Scheduling and CUDA graphs: the SM100 variable-length dynamic persistent scheduler and metadata preparation ([#2559](https://github.com/Dao-AILab/flash-attention/pull/2559)); preservation of first-tile scheduler state ([#2705](https://github.com/Dao-AILab/flash-attention/pull/2705)); the variable-length block-sparse SplitKV deadlock fix ([#2761](https://github.com/Dao-AILab/flash-attention/pull/2761)); updated CLC fuzz expectations ([#2766](https://github.com/Dao-AILab/flash-attention/pull/2766)).
- Numerical and runtime correctness: stable backward and forward compile keys for tensor-valued maximum sequence lengths ([#2507](https://github.com/Dao-AILab/flash-attention/pull/2507), [#2762](https://github.com/Dao-AILab/flash-attention/pull/2762)); forward dynamic-shape corrections ([#2745](https://github.com/Dao-AILab/flash-attention/pull/2745)); paged-KV block-table bounds validation ([#2711](https://github.com/Dao-AILab/flash-attention/pull/2711)); dtype-aware E4M3 softmax bounds and later FP8 numerical refinements ([#2717](https://github.com/Dao-AILab/flash-attention/pull/2717), [#2731](https://github.com/Dao-AILab/flash-attention/pull/2731)); sparse MLA sentinel-index protection ([#2755](https://github.com/Dao-AILab/flash-attention/pull/2755)); SM90 block-sparse backward preprocessing synchronization ([#2756](https://github.com/Dao-AILab/flash-attention/pull/2756)).
- Build and dependency maintenance: dropout-disabled header cleanup ([#2669](https://github.com/Dao-AILab/flash-attention/pull/2669)); corrected ROCm CK variable-length forward bindings ([#2742](https://github.com/Dao-AILab/flash-attention/pull/2742)); compatibility with the current Quack packed-subtraction API ([#2787](https://github.com/Dao-AILab/flash-attention/pull/2787)); a relaxed `nvidia-cutlass-dsl>=4.6.2` requirement ([#2798](https://github.com/Dao-AILab/flash-attention/pull/2798)); faster scalar SM100 mask compilation ([#2819](https://github.com/Dao-AILab/flash-attention/pull/2819)).
- Documentation and cleanup: kernel-debugging methodology documentation ([#2753](https://github.com/Dao-AILab/flash-attention/pull/2753)) and a layer-normalization comment correction ([#2744](https://github.com/Dao-AILab/flash-attention/pull/2744)).
- Intentionally excluded FA3 changes: the `hopper/` portions of upstream [#2692](https://github.com/Dao-AILab/flash-attention/pull/2692), [#2702](https://github.com/Dao-AILab/flash-attention/pull/2702), and [#2746](https://github.com/Dao-AILab/flash-attention/pull/2746) are absent from the final tree because the vLLM fork keeps FlashAttention 3 frozen.

## Downstream integration

- Kept the entire `hopper/` directory byte-for-byte identical to fork `main` (tree `f4399d1f2f57ba66d9be20c304f81697cacd84d5`), so this sync does not modify FlashAttention 3. Preserved the fork's CMake build and PyTorch stable ABI registration for FA2 and FA3, including the existing exclusion of the upstream `hopper/flash_attn_3` package.
- Retained fused output quantization, the compile API, dynamic causal input, SM90 FP8-KV dequantization, and the fork's head-dimension 256 kernels and tuning.
- Integrated the dynamic variable-length scheduler for SM100 and SM110. SM90 continues to use its fixed SplitKV scheduler; unsupported per-batch split metadata and virtual-batch reordering raise descriptive errors.
- Preserved the existing `flash_attn_combine` keyword contract while adopting upstream terminology: `virtual_batch_idx` is canonical, `varlen_batch_idx` remains accepted as a compatibility alias, positional ordering is unchanged, and supplying both names is rejected.
- Extended compile-only tensor specifications through scheduler preparation, tile-prefix-sum, and block-index helpers. Compile-only warmup populates the same cache entries used at runtime without launching those helper kernels.
- Ported the paged-KV bounds check from Dao-AILab #2711 to the stable FA2 ABI with `torch::stable::amax` and AOTI scalar extraction.
- Advanced `csrc/composable_kernel` to `c56c6750d0fc54ed771d532cc92c316423449614` and updated the CK forward wrappers with the matching signatures. `third_party/aiter` advances to `9bab8388c35936814a659b4ebd245c491e1b940a`.
- Added focused coverage for compile-only tensor alignment, scheduler cache reuse, SM90 fixed-split behavior, large variable-length batches, and paged-KV capacity checks.
- Corrected the fixed-length FA4 autograd return tuple after the downstream fused-output API added `out` and `output_scale` to `FlashAttnFunc.apply`. H100 backward coverage exposed the 21-versus-23 gradient arity mismatch; the corrected path passed fixed-length and variable-length backward tests.

## Companion vLLM validation

At Matthew Bonanni's request, the pre-landing companion is [vllm-project/vllm#54819](https://github.com/vllm-project/vllm/pull/54819). It is marked `[Do Not Merge]`, points `GIT_REPOSITORY` to the `StevenWang-CY` fork, and pins exact candidate `ee391c01f50d48a13d7e3d584f072bb0dba8dc4c` so vLLM CI can validate this sync before it lands. After #188 lands, the companion must restore the official repository URL and pin the final merged commit before the DNM marker is removed.

## Validation

Static and history checks were rerun on exact commit `ee391c01f50d48a13d7e3d584f072bb0dba8dc4c` (tree `b4c7f467f3874ca7b8fb7471abee0d363b5113ff`) against current fork `main` at `88a8b4c9980b7060ac249742b4be8dd8a2769e9d`. The two-file resolution ports the landed hd256 SplitKV clamp from #192 into this branch's refactored `_get_fwd_config` and preserves its SM100/SM110 regression test. In response to review, the final follow-up restores every `hopper/` file to fork `main`: the `hopper/` tree is exactly `f4399d1f2f57ba66d9be20c304f81697cacd84d5`, all 55 remaining PR files are outside `hopper/`, and every non-hopper byte matches the previously validated head. The H100 FA2 and FA4 results below were collected on non-hopper code that remains unchanged. The earlier FA3 measurements used the pre-freeze Hopper tree and are retained only as historical context; the final candidate does not change FA3.

### Static and history checks

- `git diff --check upstream/main...HEAD`
- `python3 -m compileall -q flash_attn hopper tests setup.py`
- `uvx pre-commit==4.4.0 run --all-files`
- The repository's latest-Ruff check and format commands from `.github/workflows/ci.yml`
- Isolated CPU contract execution for the canonical and compatibility combine keywords, preserved positional ordering, and the dual-key rejection path
- A 42-case scheduling matrix across SM100/SM110 auto, explicit, local, and packed-GQA paths; mutation controls for both missing-clamp failure modes; and AST parity with #192's landed predicate and regression test
- Conflict-marker scan over the complete tracked tree and clean-worktree verification after recursive submodule initialization
- Commit audit: all 33 commits carry DCO sign-offs, 27 include exact source references, upstream human authorship is preserved, and no commit is empty

### H100 execution

Environment: NVIDIA H100 80GB HBM3 (SM90, 132 SMs), driver 570.148.08, CUDA 12.8, Python 3.10.12, PyTorch 2.11.0+cu128, `nvidia-cutlass-dsl` 4.6.2, `quack-kernels` 0.6.4, and `apache-tvm-ffi` 0.1.11.

- On the pre-review synchronized tree at `2dc4ebe11e79ef28b95c80fb181cf69780c53e9f`, built the native C++ and CUDA tree for compute capability 9.0a and loaded `_vllm_fa2_C.abi3.so` and `_vllm_fa3_C.abi3.so`. The FA2 measurements remain applicable because its source is unchanged. The FA3 measurements predate the requested `hopper/` freeze and are historical only; the final candidate's FA3 source is exactly fork `main`.
- Stable-ABI probes passed for FA2 paged-KV bounds: cache-length and append overflow were rejected, while the exact-capacity control launched. FA2's default and explicit-zero split settings were bitwise equal, and unsupported explicit splitting was rejected. Historical pre-freeze FA3 result: split-4 matched unsplit within `0.0009765625` and an FP32 SDPA reference within `0.00102922`.
- The selected FA4 source tests ran unmodified and completed with 29 passes and 4 expected SM90 skips. Coverage included fixed-SplitKV compile and runtime parity, scheduler cache and combine paths, learnable-sink backward, cumulative-sequence metadata, block-sparse inference, score-modifier backward, and fixed-length and variable-length forward-backward controls. The skips are the SM90 cases that explicitly require variable-length SplitKV, which is not implemented for that architecture.
- Configured vLLM main at `6ec92bcbc8ef9af25cb4834ba3d922115792fbeb` with `VLLM_FLASH_ATTN_SRC_DIR` pointing to the pre-review checkout, then exercised its wrappers on one BF16 GQA variable-length case through FA2, FA3 split-1 and split-4, and FA4 split-1 and split-2. Maximum FP32 SDPA reference errors were `0.0010855198`, `0.0011264384`, and `0.0010855198`, respectively. FA3's maximum split difference was `0.0009765625`; FA4's was zero. The FA2 and FA4 portions remain applicable to the unchanged non-hopper code; the FA3 portion is historical only. The SM90 fixed-split-2 FA4 compile-from-specifications path also completed.

The exact PyTorch 2.13 and CUDA 13 runtime declared by current vLLM main, `apache-tvm-ffi>=0.1.12`, SM103, SM100, SM110, SM120, and ROCm were not exercised here and require maintainer or companion-vLLM hardware validation. The source-override integration above used the stated CUDA 12.8 environment; its `apache-tvm-ffi` 0.1.11 is below the package's declared `>=0.1.12` floor.

### CODE REVIEW

**Comment by StevenWang-CY** [comment]
@MatthewBonanni The upstream sync you suggested is ready for review. I completed an H100 build, focused FA2/FA3/FA4 validation, and a current-vLLM source-override probe; the exact results are in the description. The H100 run also exposed an FA4 autograd return-arity mismatch, which is fixed in the final commit. After this lands, I will submit the companion vLLM `GIT_TAG` update following vllm-project/vllm#41052.

**Comment by StevenWang-CY** [comment]
@MatthewBonanni, could you review and approve this upstream sync when convenient? I revalidated the exact head 2802a0f against the current fork and Dao main: both bases are unchanged, all 26 Dao commits are present in order, the 31-commit DCO and provenance audit is clean, and compileall, current Ruff, pre-commit, positional and keyword compatibility, and autograd arity checks pass. The documented H100 build, FA2, FA3, FA4, and vLLM source-override results remain applicable. I will address any findings promptly.

**Comment by MatthewBonanni** [comment]
Thanks for the PR! Sorry I haven't had a chance to look at it yet. Please submit the vLLM-side PR first so we can run vLLM CI before landing this. Mark it as [Do Not Merge] in the title and change the GIT_REPOSITORY to your fork

**Comment by StevenWang-CY** [comment]
Submitted as requested: [vllm-project/vllm#54819](https://github.com/vllm-project/vllm/pull/54819). It is marked `[Do Not Merge]`, uses `https://github.com/StevenWang-CY/flash-attention.git`, and pins exact candidate `e7a154ff85b0aa793be925fbe97df6b4e5f26ce7`. I also updated this PR's description to reflect the pre-landing validation sequence.

The vLLM pre-run gate currently awaits repository authorization; DCO and the other early metadata checks are green. I will follow the CI results and, after this sync lands, restore the official repository URL and update the companion pin to the final merged commit before removing the DNM marker.

File: hopper/flash_fwd_combine_kernel.h [hunk 1]
```diff
@@ -215,6 +215,7 @@ class FlashAttnFwdCombine {
         int const* cu_seqlens_q;
         int const* seqused_q;
         int const* prepare_seqlen_q_ptr;
+        int const* varlen_batch_idx_ptr;
```
Comment (MatthewBonanni) [inline hopper/flash_fwd_combine_kernel.h]: Please revert all changes to FlashAttention 3 (`hopper` folder); we're leaving this frozen in our fork

**Review by MatthewBonanni — COMMENTED** [review]
CI is looking good, thanks! One quick comment

File: hopper/flash_fwd_combine_kernel.h (hunk 1 above)
Comment (StevenWang-CY) [reply hopper/flash_fwd_combine_kernel.h]: Done in `ee391c01f50d48a13d7e3d584f072bb0dba8dc4c`. The complete `hopper/` subtree now exactly matches current fork `main` (tree `f4399d1f2f57ba66d9be20c304f81697cacd84d5`), and no `hopper/` path remains in the PR diff. I also updated vllm-project/vllm#54819 to pin this exact head and requested fresh CI.

Comment (StevenWang-CY) [reply hopper/flash_fwd_combine_kernel.h]: Follow-up validation is now settled. On exact companion head 9d8ef1449e, GitHub shows 293 successful, 3 skipped, and 1 cancelled contexts, with none pending or failing. The lone Buildkite soft failure is the same Ascend image setup issue seen twice in #86791 (triton-ascend==3.2.2 unavailable), before any tests and unrelated to the one-file pin; all three retried test jobs passed. The hopper/ subtree here remains byte-identical to main. When convenient, could you please take another look?

**Review by MatthewBonanni — APPROVED** [review]
LGTM, thanks for doing this!
